### DISCUSSION TITLE
Support for bl0942 energy monitoring

### DISCUSSION THREAD

**Discussion by GreJ0** [arendst/Tasmota#13190] [Ideas]
I have a power plug with IC bl0942 for energy monitoring. Is it possible to get this driver in Tasmota like the BL0940?

**Comment by barbudor** [comment]
That's Shangai Belling week !
After the BL0939 in the Sonoff Dual R3 v2, now a power plug with BL042

@GreJ0 we would need good photos and some tracking of the pins on the PCB
Is that ESP8266 or ESP32 ?

Datasheets are available https://www.belling.com.cn/product_info.html?id=370
But I'm already on the BL0939
@arendst  ?

**Reply by barbudor** [reply]
@GreJ0 Have you already tasmotized the device ? If yes, did you first kept a copy of the original firmware ?

**Reply by barbudor** [reply]
@GreJ0 Have you tried the BL040 driver ? does it provide results ? may be wrong, but would be interesting to know if it the protocol is the same.

**Reply by GreJ0** [reply]
The device had a CB2S Module (Tuya) I replaced it with a TYWE2S Module with Tasmota. So I don't have the original firmware. I have try the BL0940 but no result.

**Comment by GreJ0** [comment]
The device had a CB2S Module (Tuya) I replaced it with a TYWE2S Module with Tasmota. So I don't have the original firmware. I have try the BL0940 but no result. 
![20210918_093922](https://user-images.githubusercontent.com/56791933/134048239-e7f5e165-8d6d-4564-b82a-f8c8d7de9809.jpg)
![20210918_003054](https://user-images.githubusercontent.com/56791933/134048307-177c8296-f4e8-43a4-b71c-aa0228d5b750.jpg)
Protocol different?
![image](https://user-images.githubusercontent.com/56791933/134048558-4ea1c5ce-f5c9-421f-9eac-522d490a02b4.png)
![image](https://user-images.githubusercontent.com/56791933/134048836-c3c7432c-ab0b-4920-8bcf-4496f9c0b2f9.png)

**Reply by barbudor** [reply]
Thanks, looking further in the datasheet the data packet is actually different between 942 and 940 as you pointed
but that should not be much different in the principle
I'll to finish the 939 first and then I'll have a look at the 942

**Reply by GreJ0** [reply]
ok thanks in advance

**Reply by barbudor** [reply]
@GreJ0 
Do you still have the original module ?
Would it be possible to power-it (3.3V) and see if you get serial data on the TX pin ?
You can use the same setup that you used to flash the ESP module (same serial adapter)
But it would require a terminal software that can show HEX data (like RealTerm http://www.i2cchip.com/realterm)
Baudrate is likely to be 4800 or 9600

If you can find a patterns with a lot of 0x5*, 0xA*, 0xAA that is probably the right thing

Thanks

**Reply by barbudor** [reply]
@GreJ0 
There are 2 things we need to identify:
- The baudrate (4800 or 9600)
- The address (there is a version of the chip that support multi-device on the same serial), the default should be 0 but we have already seen info in the datasheet not being correct).

Could you please help with the following:
1st step, you need to assign `SerBr TX` to the GPIO connected to the BL0942 RX/SDI (pin 9) and `SerBr RX` to the GPIO connected to the BL0942 TX/SDO (pin 10).
After reboot, configure the baudrate `SBaudrate 4800`
You need to send the hex string that requests the data to the BL0942
Try `SSerialSend5  58AA`. It should returns a string of 22 bytes starting with `55`.
If nothing is returned, try with `SSerialSend5  59AA`, `SSerialSend5  5AAA` or `SSerialSend5  5BAA`
If you get nothing, the baudrate is probably wrong
Change to `SBaudrate 9600` and re-issue the same commands until you get a response.

Thanks

**Reply by GreJ0** [reply]
Should I do this in the Tasmota consol? And is GPIO1 TX and GPIO3 RX?
I don't get any reaction, can you advice me?
I only found a string of 58AA with Realterm. and the baudrate is 4800

**Reply by barbudor** [reply]
Hi @GreJ0 
Thanks. So your test with the original module confirmed the baudrate to be 4800 and also the subaddress of the module to be 0 (the last 2 bits of 0x58)
If you have still the module around, I would be interested too in the the first 2 seconds of communication from the module. Before sending those `58 AA` the module is probably sending some other configuration strings.

Regarding the other questions, yes, with GPIO 1 & 3 configured accordingly as TX and RX, `SBaudrate 4800`, you should be able to send yourself `SSerialSend5 58AA` and the BL042 should reply with a a packet of 22 bytes that I'm interested in to confront them to the datasheet
Be sure to use ` SSerialSend5`  and not `SSerialSend`. The 5 is important to tell Tasmota that the string is in hexa
I can be reached on Discord too if you want

**Comment by GreJ0** [comment]
@barbudor
I shall try it this weekend.

Op ma 20 sep. 2021 22:40 schreef Barbudor ***@***.***>:

**Comment by GreJ0** [comment]
@barbudor,

I found this, is this what you mean?

![image](https://user-images.githubusercontent.com/56791933/134904269-9a94e7b8-f7f2-4c11-b53b-889421b28d3a.png)

Op ma 20 sep. 2021 om 22:40 schreef Barbudor ***@***.***>:

> @GreJ0 <https://github.com/GreJ0>
> Do you still have the original module ?
> Would it be possible to power-it (3.3V) and see if you get serial data on
> the TX pin ?
> You can use the same setup that you used to flash the ESP module (same
> serial adapter)
> But it would require a terminal software that can show HEX data (like
> RealTerm http://www.i2cchip.com/realterm)
> Baudrate is likely to be 4800 or 9600
>
> If you can find a patterns with a lot of 0x5*, 0xA*, 0xAA that is probably
> the right thing
>
> Thanks
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/arendst/Tasmota/discussions/13190#discussioncomment-1359346>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANRJG7IAMCFZPVO52F2EQSTUC6L23ANCNFSM5EMI42HQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 

*Grt,*

*John*
